#### No7 (3 балла)
С сайта https://www.kaggle.com/ скачайте датасет на ваш выбор для обучения модели
логистической регрессии (задача классификации). Выполните следующие действия:

In [1]:
import pandas as pd


In [2]:
df = pd.read_csv(
    "csv/heart.csv").drop(["oldpeak", "slp", "caa", "thall"], axis=1)
df.head()


age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  output
0   63    1   3     145   233    1        0       150     0       1
1   37    1   2     130   250    0        1       187     0       1
2   41    0   1     130   204    0        0       172     0       1
3   56    1   1     120   236    0        1       178     0       1
4   57    0   0     120   354    0        1       163     1       1

1) Опишите выбранный набор данных: контекст данных, какие данные есть в наборе,
как можно будет использовать полученные результаты классификации.

Датасет отображает данные о признаках сердечного приступа

- age - возраст	
- sex - пол
- cp - тип грудной боли
- trtbps - давление в состоянии покоя
- chol - холестерин в мг/дл, полученный с помощью датчика ИМТ
- fbs - уровень сахара в крови натощак > 120 мг/дл) (1 = да; 0 = нет)
- restecg - результаты электрокардиографии в состоянии покоя
- thalachh - максимальный пульс
- exng - стенокардия, вызванная физической нагрузкой (1 = да; 0 = нет)
- output - исход

Данные можно будет использовать для предсказания сердечного приступа

In [3]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   age       303 non-null    int64
 1   sex       303 non-null    int64
 2   cp        303 non-null    int64
 3   trtbps    303 non-null    int64
 4   chol      303 non-null    int64
 5   fbs       303 non-null    int64
 6   restecg   303 non-null    int64
 7   thalachh  303 non-null    int64
 8   exng      303 non-null    int64
 9   output    303 non-null    int64
dtypes: int64(10)
memory usage: 23.8 KB


In [4]:
df.corr()


age       sex        cp    trtbps      chol       fbs  \
age       1.000000 -0.098447 -0.068653  0.279351  0.213678  0.121308   
sex      -0.098447  1.000000 -0.049353 -0.056769 -0.197912  0.045032   
cp       -0.068653 -0.049353  1.000000  0.047608 -0.076904  0.094444   
trtbps    0.279351 -0.056769  0.047608  1.000000  0.123174  0.177531   
chol      0.213678 -0.197912 -0.076904  0.123174  1.000000  0.013294   
fbs       0.121308  0.045032  0.094444  0.177531  0.013294  1.000000   
restecg  -0.116211 -0.058196  0.044421 -0.114103 -0.151040 -0.084189   
thalachh -0.398522 -0.044020  0.295762 -0.046698 -0.009940 -0.008567   
exng      0.096801  0.141664 -0.394280  0.067616  0.067023  0.025665   
output   -0.225439 -0.280937  0.433798 -0.144931 -0.085239 -0.028046   

           restecg  thalachh      exng    output  
age      -0.116211 -0.398522  0.096801 -0.225439  
sex      -0.058196 -0.044020  0.141664 -0.280937  
cp        0.044421  0.295762 -0.394280  0.433798  
trtbps   -0.114103 -0.046698  0.067616 -0.144931  
chol     -0.151040 -0.009940  0.067023 -0.085239  
fbs      -0.084189 -0.008567  0.025665 -0.028046  
restecg   1.000000  0.044123 -0.070733  0.137230  
thalachh  0.044123  1.000000 -0.378812  0.421741  
exng     -0.070733 -0.378812  1.000000 -0.436757  
output    0.137230  0.421741 -0.436757  1.000000

2) Постройте классификацию, используя логистическую регрессионную модель.
Оцените качество работы модели.

In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

x = df.drop("output", axis=1)
y = df["output"]
x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size=0.3, random_state=42)

lr = LogisticRegression(max_iter=500)
lr.fit(x_train, y_train)
lr.score(x_test, y_test)


0.8021978021978022

3) Спрогнозируйте несколько новых значений с помощью построенной модели.

In [6]:
# 1
data1 = df.iloc[45].drop("output")
data1

# 0
data2 = df.iloc[165].drop("output")
data2

test_df = pd.DataFrame([data1, data2])
test_df


age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng
45    52    1   1     120   325    0        1       172     0
165   67    1   0     160   286    0        0       108     1

In [7]:
test_df["output"] = lr.predict(test_df)
test_df


age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  output
45    52    1   1     120   325    0        1       172     0       1
165   67    1   0     160   286    0        0       108     1       0